test/versioned/winston/winston.tap.js

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const tap = require('tap')
+const helper = require('../../lib/agent_helper')
+const concat = require('concat-stream')
+
+tap.test('Winston instrumentation', (t) => {
+  t.autoend()
+
+  let agent
+  let winston
+
+  function setup(config) {
+    agent = helper.instrumentMockedAgent(config)
+    winston = require('winston')
+  }
+
+  t.afterEach(() => {
+    agent && helper.unloadAgent(agent)
+    winston = null
+    // must purge require cache of winston related instrumentation
+    // otherwise it will not re-register on subsequent test runs
+    Object.keys(require.cache).forEach((key) => {
+      if (/winston/.test(key)) {
+        delete require.cache[key]
+      }
+    })
+  })
+
+  // Stream factory for a test. Applies common assertions to logged messages.
+  const makeStreamTest = (cb) => {
+    let toBeClosed = 0
+    return (assertFn) => {
+      toBeClosed++
+      return (msgs) => {
+        for (const msg of msgs) {
+          assertFn(msg)
+        }
+        if (--toBeClosed === 0) {
+          cb()
+        }
+      }
+    }
+  }
+
+  const logStuff = (logger, streams) => {
+    // Log some stuff, both in and out of a transaction
+    logger.info('out of trans')
```

Comment:
Since we do some things with log levels, although admittedly not in this test, should we also do some testing at various levels? For example, that we really don't log anything at the `debug` level because the logger was configured at the `info` level. So test that it really is so?

Reply:
I would be inclined to include those tests in the log enricher itself, like [here](https://github.com/newrelic/newrelic-node-log-extensions/blob/main/packages/winston-log-enricher/tests/versioned/winston.tap.js) rather than in this auto-instrumentation, where we are really just checking that we're able to hook onto the module loading to apply the log enricher. What do you think?

Reply:
Good point. I'd make a case we should pare down these tests to avoid duplication in log enricher. But just enough to provide the log enricher can be instrumented and works as designed. Basically, how I found the problem with incorrect Winston instance getting enriched.

Reply:
I've updated the test suite to focus on the specific code paths in our instrumentation, rather than the formatter.